pkg/container/versioned/value.go

```diff
@@ -0,0 +1,337 @@
+// Value is a container for versioned value, implemented as a lock-free linked list
+type Value[T any] struct {
+	// versionedNodes are non-overlapping and in sorted order
+	head atomic.Pointer[valueNode[T]]
+}
+
+// SetValueAtVersion adds the value with validity starting from 'version'.
+// All values are added with "infinite" validity, which is then truncated
+// when an overlapping entry is added, or a value is removed.
+//
+// 'version' must be later than any the current version visible to the readers.
+// Must be called from a single writer!
+func SetValueAtVersion[T any](v *Value[T], value T, version Version) {
+	newNode := &valueNode[T]{
+		versions: versionRange{
+			first: version,
+		},
+		value: value,
+	}
+	newNode.versions.past.Store(invalidVersion)
+
+	// find if there is a current value that is valid for this new version
+	node := v.head.Load()
+	prev := &v.head
+	for node != nil {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			log.Errorf("AddValueAtVersion may not add values at versions lower than those already added! (%d < %d). Stacktrace: %s", version, node.versions.first, stacktrace)
+		}
+
+		if node.versions.Contains(version) {
+			// truncate the validity of this node to end at 'version'
+			node.versions.past.Store(version)
+			// link the new node after the current one
+			newNode.next.Store(node.next.Load())
+			node.next.Store(newNode)
+			break
+		}
+
+		node = node.next.Load()
+		prev = &node.next
+	}
+	if node == nil {
+		// Add the new value at the end
+		prev.Store(newNode)
+	}
+}
+
+// RemoveValueAtVersion changes the validity of the stored value valid at 'version' to have ended at
+// 'version'.
+// 'version' must be later than any the current version visible to the readers.
+// Must be called from a single writer!
+func RemoveValueAtVersion[T any](v *Value[T], version Version) {
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			log.Errorf("DeleteValueAtVersion may not be called with version lower than existing already! (%d < %d). Stacktrace: %s", version, node.versions.first, stacktrace)
+		}
+
+		if node.versions.Contains(version) {
+			// truncate the validity of this node to end at 'version'
+			node.versions.past.Store(version)
+			break
+		}
+	}
+}
+
+// Cleaner removes all values whose validity ends at or before 'version'.
+//
+// Must be called from a single writer!
+func Cleaner[T any](v *Value[T], keepVersion Version) {
+	// find all values that are no longer valid at 'version'
+	node := v.head.Load()
+	for node != nil && node.versions.past.Load() <= keepVersion {
+		// This node is no longer visible for readers who are all version 'keepVersion' or
+		// later.
+		node = node.next.Load()
+		v.head.Store(node)
+	}
+}
+
+// GetValue returns value of type 'T' valid for the given handle, or an empty value if none is
+// found.
+func GetValue[T any](v *Value[T], handle Handle) T {
```

Comment:
This API reads a bit strange to me, possibly because it's ported from C? In Go, I'd expect something like

`val := value.At(version)` instead of `val := GetValue(value, version)`

To me, just `Handle` is also not descriptive enough, maybe `VersionRef` or `VersionHandle` would communicate more clearly that this is a reference to a version. Alternatively, we could make a current `Version` be `VersionNum` or a `Revision` and then have the `Handle` be the `Version`, but I'm not sure that's clearer.

Reply:
Renamed `Handle` as `VersionHold` and made it to accessed via a pointer. The value of `VersionHold` is not to be copied around, as that would mess with tracking of stale values. Also added a finalizer to catch un-closed `VersionHolds()`.

Reply:
I'm a bit confused why we can't do `value.At`. We are not introducing new type parameters, so this type should be fine:
``
func (v *Value[T]) At(h VersionHold) T { ... }
```

Reply:
Actually all the functions that take Value[T] should just be normal methods on it.

Reply:
Yes, thanks for pointing this out. I'm learning Go generics "on-the-go", and faced a compiler error relating to something like this. It must have been that the method in question introduced a new generic type, and I thought the error to mean that no generic type can be a receiver...

Reply:
Yeah, here's the reasoning - essentially you can't _add_ more type parameters in a method, but the method can use the type parameters of its receiver.
https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#no-parameterized-methods